mod/calculated_metrics/spec/lib/formula/calculator/input_item/company_option/company_query_spec.rb

```diff
@@ -0,0 +1,218 @@
+        .to eq_sql <<-SQL
+                   SELECT r0.subject_company_id, r0.year, r0.object_company_id
+                   FROM relationships AS r0
+                   LEFT JOIN relationships AS r1 ON
+                   r0.subject_company_id = r1.subject_company_id &&
+                   r0.object_company_id = r1.object_company_id && r0.year = r1.year
+                   WHERE ((r0.metric_id = #{more_evil_id} && r0.value = "yes") AND
+                   (r1.metric_id = #{supplied_by_id} && r1.value = "Tier 1"))
+      SQL
+    end
+
+    example "or condition" do
+      expect(sql("Related[Jedi+more evil = yes || Commons+supplied_by=Tier 1]"))
+        .to eq_sql <<-SQL
+                   SELECT r0.subject_company_id, r0.year, r0.object_company_id
+                   FROM relationships AS r0
+                   WHERE ((r0.metric_id = #{more_evil_id} && r0.value = "yes") OR
+                   (r0.metric_id = #{supplied_by_id} && r0.value = "Tier 1"))
+      SQL
+    end
+
+    example "with answer condition" do
+      expect(sql("Related[Jedi+more evil = yes] && Jedi+deadliness > 10"))
+        .to eq_sql <<-SQL
+                   SELECT r0.subject_company_id, r0.year, r0.object_company_id
+                   FROM relationships AS r0
+                   LEFT JOIN answers AS a0 ON
+                   r0.object_company_id = a0.company_id && r0.year = a0.year
+                   WHERE ((r0.metric_id = #{more_evil_id} && r0.value = "yes")) &&
+                   ((a0.metric_id = #{deadliness_id} && a0.numeric_value > 10))
+      SQL
+    end
+
+    example "inverse metric" do
+      expect(sql("Related[Jedi+less evil]"))
+        .to eq_sql <<-SQL
+                   SELECT r0.object_company_id, r0.year, r0.subject_company_id
+                   FROM relationships AS r0
+                   WHERE ((r0.inverse_metric_id = #{less_evil_id}))
+      SQL
+    end
+
+    example "inverse metric with or condition" do
+      expect(sql("Related[Jedi+less evil || Commons+supplied_by=Tier 1]"))
+        .to eq_sql <<-SQL
+                   SELECT r0.subject_company_id, r0.year, r0.object_company_id
+                   FROM (
+                     SELECT subject_company_id, object_company_id,
+                            metric_id, inverse_metric_id, year, value
+                     FROM relationships
+                     WHERE metric_id = #{supplied_by_id}
+                     UNION
+                     SELECT object_company_id as subject_company_id,
+                            subject_company_id as object_company_id,
+                            metric_id, inverse_metric_id, year, value
+                     FROM relationships
+                     WHERE inverse_metric_id = #{less_evil_id}
+                   ) AS r0
+                   WHERE ((r0.inverse_metric_id = #{less_evil_id})
+                   OR (r0.metric_id = #{supplied_by_id} && r0.value = "Tier 1"))
+      SQL
```

Comment:
I'm not usually in favor of long exact string matching in these, but I agree that just testing results doesn't capture enough here.